Add "Mod Weapon Vars"

EladNLG opened the pull request:
## What?

Creates an alternative to weapon mods by allowing overrides of any _numeric_ eWeaponVar through script.

### Example

`ScriptWeaponVars_SetInt(weapon, eWeaponVar.ammo_clip_size, 50)`

## Why?

1. Weapon files need to be downloaded client side to work, and MAD isn't finished yet.
2. Weapon mods have a 200 "modifications" limit. This has been raised to ~240 through admirable work of others, but will not be enough for larger-scale mods/mods that want to apply a variety of modifiers on a variety of weapons.
3. Allows setting a variable to a range of values instead of being limited to 32 options (e.g. if you'd want to recreate the nemesis with the current weapon system, you wouldn't be able to because the amount of possible values for its burst delay is effectively infinite, however, this allows such things.)

WIP. Relies on

- #692

### Required before merging
- [x] Manager for script weapon modifiers
- [x] System to reapply overrides when weapon vars are recalculated (e.g. after `AddMod`) (Script PR not existing yet)

### Optional
~~- [ ] Auto-sync server changes to client through a remote function~~
~~- [ ] Add "predicted" versions of functions, that do not transmit to client~~
Decided against autotransmitting changes for a few reasons:
1) MAD
2) would be a pain and overcomplicate things imo

<!-- 
WHEN OPENING A PULL REQUEST KEEP IN MIND:
-> If the changes you made can be summarised in a screenshot, add one (e.g. you changed the layout of an in-game menu)
-> If the changes you made can be summarised in a screenrecording, add one (e.g. proof that you fixed a certain bug)

-> For fixes, description on how to reproduce the bug are appreciated and help your PR get merged faster
-> For features, description on how to use the feature is appreciated and will help your PR get merged faster

-> Please use a sensible title for your pull request

-> Please describe the changes you made. The easier it is to understand what you changed, the higher the chances of your PR being merged (in a timely manner).

-> If you made multiple independent changes, please make a new PR for each one. This prevents your PR being blocked from merging by one of the changes you made.

Note that commit messages in PRs will generally be squashed to keep commit history clean.
-->

\[Screenshots/Examples TBD\]

EladNLG committed:
Add mod weapon vars

create an alternative to weapon mods by allowing overrides of any numeric eWeaponVar.

EladNLG committed:
Cleanup code

EM4Volts commented:
while im not able to "review" this because i dont know anything in c++ or such at all i just want to say that his would be an excelent addition, the mod limit is big but still something that can happen. and having the evars modifyable directly via script enables alot more then the current need of switching mods on stuff.

EladNLG committed:
Replace magic offset with structs

ASpoonPlaysGames commented on primedev/client/weaponx.cpp:
If there isn't anything in here, not much point in it existing

EladNLG replied:
kinda dont like having a header without a source counterpart, but sure

ASpoonPlaysGames commented on primedev/mods/modweaponvars.cpp:
```diff
@@ -0,0 +1,103 @@
+// rn we are "disabling" this function because it seems to make stuff
+// work, figure out why.
+// also ""slight"" performance increase
+AUTOHOOK(Cl_UpdateWeaponVars, client.dll + 0x5B2670, char, __fastcall, (void* a2))
```
Bit scary but sure

EladNLG replied:
A better solution is to probably hook the function that actually recalculates mod values and do a codecallback after it

ASpoonPlaysGames commented on primedev/mods/modweaponvars.h:
Do you actually need to cast these? wont the compiler implicitly convert them?

ASpoonPlaysGames commented on primedev/server/weaponx.h:
```diff
@@ -0,0 +1,14 @@
+	// this is used by the game as the initial offset
+	// for getting eWeaponVar values. No idea about
+	// the actual datatype. Start of another struct?
+	const char weaponVars[0x1000];
```
Does native do any checks to ensure that this is within some bounds? if so, you could probably figure out the exact size

EladNLG replied:
could probably get the highest offset and then add 8 bytes to that, should be enough for our use case?

catornot commented on primedev/mods/modweaponvars.cpp:
```diff
@@ -0,0 +1,103 @@
+
+	if (context == ScriptContext::SERVER)
+	{
+		CWeaponX* weapon = (CWeaponX*)ent;
+		*(float*)(&weapon->weaponVars[varInfo->offset]) = value;
```
should add a check here for if it's a weapon entity

catornot commented on primedev/mods/modweaponvars.cpp:
```diff
@@ -0,0 +1,103 @@
+	}
+	else // if (context == ScriptContext::CLIENT)
+	{
+		C_WeaponX* weapon = (C_WeaponX*)ent;
+		*(float*)(&weapon->weaponVars[varInfo->offset]) = value;
```
same here

EladNLG committed:
Add type checking

EladNLG committed:
Remove unnecessary files

EladNLG committed:
Several Changes (See description)

- reformat some error messages
- add support for boolean values
- expand WeaponVarType enum
- set weaponVars length to the correct length
- rename functions
- replace hook for random func with hooks to Cl_CalcWeaponMods and Sv_CalcWeaponMods

EladNLG committed:
Add CreateScriptInstance

EladNLG committed:
Format fixes, remove AddCallback_ApplyModWeaponVars

we will be doing a script PR instead, as it is easier

EladNLG committed:
Format Fix

EladNLG committed:
Remove unrelated, not-working code

EladNLG committed:
Add codecallback call

EladNLG committed:
Fix SQMessageBufferPushArg

EladNLG committed:
Add RecalculateModsForWeapon

EladNLG committed:
Add client-side RecalculateModsForWeapon

EladNLG committed:
Fix prediction using unholy magic

EladNLG commented:
PR finally works with prediction - gonna try cleaning up code tomorrow.

_Unfortunately_, i cannot figure out proper naming for the hooked functions - im just going off of frequency of calls :[

Please help if you can.

EladNLG committed:
Remove unnecessary hook, rename "thingy" to Cl_WeaponTick

i think thats what it is? makes the most sense to me
also format fix

EladNLG committed:
Format fix

EladNLG commented:
Also, do we want to support strings/assets/vectors?

ASpoonPlaysGames commented:
> Also, do we want to support strings/assets/vectors?

If it's not that bad, yeah would be nice. Vectors probably the most useful

EladNLG committed:
Add ModWeaponVars_SetString, ModWeaponVars_GetType

also change some comments

EladNLG committed:
remove unnecessary comments

EladNLG commented on primedev/mods/modweaponvars.cpp:
```diff
@@ -0,0 +1,255 @@
+AUTOHOOK_INIT()
+
+// RENAME THIS BEFORE MERGE (REVIEWERS IF YOU STILL SEE THIS BLOCK IT NOWNOWNOWNOWNOW)
+AUTOHOOK(Cl_WeaponTick, client.dll + 0x59D1E0, bool, __fastcall, (C_WeaponX * weapon))
```
yeah, since we got rid of the other hook (we no longer need it) i can just call this Cl_WeaponTick...

the name might be wrong tho? ill be adding a comment for now

catornot commented on primedev/mods/modweaponvars.cpp:
```diff
@@ -0,0 +1,311 @@
+	return SQRESULT_NULL;
+}
+
+ADD_SQFUNC("void", ModWeaponVars_GetType, "int eWeaponVar", "", ScriptContext::SERVER | ScriptContext::CLIENT)
```
this sqfunction should be marked with a int return type probably.

EladNLG replied:
oop my mistake, will fix tomorrow

catornot commented on primedev/mods/modweaponvars.cpp:
```diff
@@ -0,0 +1,311 @@
+{
+	bool result = Sv_CalcWeaponMods(unk_0, unk_1, unk_2, unk_3, unk_4);
+
+	if (result && IsWeapon<ScriptContext::SERVER>((void**)(unk_2 - 0x1410)))
+	{
+		SQObject* entInstance = g_pSquirrel<ScriptContext::SERVER>->__sq_createscriptinstance((void**)(unk_2 - 0x1410));
```
these offsets into `unk_2` seam weird but they should still be a struct somehow imo.

catornot replied:
atleast put the offset into a named constant.
or add the field to the `CWeaponX` class and get it's offset from that.

EladNLG replied:
Its already in the weapon structs, will convert into an `offsetof` tomorrow

EladNLG committed:
Replace offsets with `offsetof`, fix GetType returntype

EladNLG committed:
rename unk_2, replace magic offsets, fix rare crash

- the rare crash would happen when the given weaponVars pointer was not from a weapon, and the location before it was not meant to be read

EladNLG committed:
Fix that crash but for real

isbadreadptr sucks actually

EladNLG committed:
Merge branch 'main' into main

EladNLG committed:
Rename function for consistency

EladNLG committed:
Make string modweaponvars work, add static asserts

GeckoEidechse committed:
Merge branch 'main' into main

GeckoEidechse commented:
Uh, for some reason format check is failing now @EladNLG :c

GeckoEidechse committed:
Merge branch 'main' into main

EladNLG commented:
Need to change hook, currently fucks up stuff (causes weird MGL-related crash, probably fucked up params?)

EladNLG committed:
Use better hooks, make naming consistent

EladNLG committed:
Merge branch 'main' into main

ASpoonPlaysGames commented:
@EladNLG Ik it's been a while, has the stuff you've been working on moved beyond this PR? Is it still worth merging?

EladNLG commented:
@ASpoonPlaysGames if people want this feature ill try to get back and work on it, but otherwise yeah im focused on other stuff now

Alystrasz committed:
Merge branch 'main' into main
